app/sprinkles/account/src/Authenticate/PasswordSecurity.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * UserFrosting (http://www.userfrosting.com)
+ *
+ * @link      https://github.com/userfrosting/UserFrosting
+ * @copyright Copyright (c) 2019 Alexander Weissman
+ * @license   https://github.com/userfrosting/UserFrosting/blob/master/LICENSE.md (MIT License)
+ */
+
+namespace UserFrosting\Sprinkle\Account\Authenticate;
+
+use UserFrosting\Sprinkle\Core\Facades\Cache;
+
+/**
+ * Handles enhanced password security methods for integration with Have I Been pwnedpasswords
+ *
+ * @see https://haveibeenpwned.com/API/v2
+ * @author Amos Folz
+ */
+class PasswordSecurity
+{
+    /**
+     * Check a password SHA1 hash against an array of compromised password hashes.
+     *
+     * @param  string $hash  The hash of the potential password to be used.
+     * @param  array  $array Array of password hashes in the format c2d18a7d49b0d4260769eb03d027066d29a:181 - or <hash>:<number of breaches.
+     * @return string $result The number of breaches password has been exposed in.
+     */
+    private static function checkHash($hash, $array)
+    {
+        foreach ($array as $index => $pwHash) {
+            $breachedItemParts = explode(':', $pwHash);
+
+            $breachedItemHash = $breachedItemParts[0];
+            $numberOfBreaches = $breachedItemParts[1];
+
+            // compare the hash suffix from Have I Been Pwned with password hash suffix.
+            if ($breachedItemHash == substr($hash, 5)) {
+                // if a match is found just return the response.
+                return $result = trim($numberOfBreaches);
+            } else {
+                $result = '0';
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Generates a list of compromised passwords stored in cache or by querying Have I Been Pwned API.
+     *
+     * First check the cache to see if the hash prefix is stored.
+     * If not found in cache, query Have I Been Pwned API and store response in cache.
+     * @see https://haveibeenpwned.com/API/v2
+     * @param  string $password
+     * @return array  $result An array containing the password checked and the number of breaches.
```

Comment:
Return doesn't need or care about the variable that is returned, just the type and any documentation. Some documentation around the data shape would also be appreciated. Having been dealing with deeply nested associated arrays for the past few months, it would be awesome if we could come up with a way to avoid it if possible.

Really wish phpDoc supported the concept of a tuple here. But alas, what could have been. https://github.com/phpDocumentor/fig-standards/pull/115/files

Potential value of using something like a class however is that the shape is inherently documented and there is much greater potential for later expansion. Probably a good idea to take a step back here and consider the best way to handle this, as I feel the helper classes could ultimately be integrated into this theoretical result class. Would be a good idea to have more discussion here I think.

Reply:
> Having been dealing with deeply nested associated arrays for the past few months, it would be awesome if we could come up with a way to avoid it if possible.

Now that you mention it, there is not really any need to be returning an array. We can simply return a string with the number of breaches. The calling method already has the password being checked so there is no reason to be passing it back and forth.